File: src/main/java/com/github/MrRogerHuang/GradleProjectInfoPlugin.java
package com.github.MrRogerHuang;

import org.gradle.api.Plugin;
import org.gradle.api.Project;
import org.gradle.api.plugins.JavaPluginConvention;
import org.gradle.api.tasks.SourceSet;
import org.gradle.tooling.provider.model.ToolingModelBuilder;
import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry;

import javax.inject.Inject;
import java.util.ArrayList;
import java.util.List;

public class GradleProjectInfoPlugin implements Plugin<Project> {
    private final ToolingModelBuilderRegistry registry;

    @Inject
    public GradleProjectInfoPlugin(ToolingModelBuilderRegistry registry) {
        this.registry = registry;
    }

    @Override
    public void apply(Project project) {
        registry.register(new CustomToolingModelBuilder());
    }

    private static class CustomToolingModelBuilder implements ToolingModelBuilder {
        @Override
        public boolean canBuild(String modelName) {
            return modelName.equals(GradleProjectInfo.class.getName());
        }

        @Override
        public Object buildAll(String modelName, Project project) {
            List<String> pluginClassNames = new ArrayList<String>();

            for(Plugin plugin : project.getPlugins()) {
            pluginClassNames.add(plugin.getClass().getName());
        }

            List<String> classpaths = new ArrayList<>();
            project.getConvention().getPlugin(JavaPluginConvention.class).getSourceSets().findByName(SourceSet.MAIN_SOURCE_SET_NAME)
                    .getRuntimeClasspath().forEach(file -> classpaths.add(file.getAbsolutePath()));


            return new DefaultModel(pluginClassNames, classpaths);
        }
    }
}

File: src/main/java/com/github/MrRogerHuang/GradleProjectInfoPlugin.kt
package com.github.MrRogerHuang

import org.gradle.api.Plugin
import org.gradle.api.Project
import org.gradle.api.plugins.JavaPluginConvention
import org.gradle.api.tasks.SourceSet
import org.gradle.tooling.provider.model.ToolingModelBuilder
import org.gradle.tooling.provider.model.ToolingModelBuilderRegistry
import java.io.File
import java.util.*
import java.util.function.Consumer
import javax.inject.Inject

class GradleProjectInfoPlugin @Inject constructor(private val registry: ToolingModelBuilderRegistry) : Plugin<Project?> {
    override fun apply(project: Project?) {
        registry.register(CustomToolingModelBuilder())
    }

    private class CustomToolingModelBuilder : ToolingModelBuilder {
        override fun canBuild(modelName: String): Boolean {
            return modelName == GradleProjectInfo::class.java.name
        }

        override fun buildAll(modelName: String, project: Project): Any {
            val pluginClassNames: MutableList<String> = ArrayList()
            for (plugin in project.plugins) {
                pluginClassNames.add(plugin.javaClass.name)
            }
            val classpaths: MutableList<String> = ArrayList()
            project.convention.getPlugin(JavaPluginConvention::class.java).sourceSets
                    .findByName(SourceSet.MAIN_SOURCE_SET_NAME)?.runtimeClasspath?.forEach(Consumer {
                        file: File -> classpaths.add(file.absolutePath)
                    })
            return DefaultModel(pluginClassNames, classpaths)
        }
    }

}